src/cdf.rs

```diff
@@ -133,18 +130,37 @@ pub struct CdfModeContext {
 }
 
 #[derive(Clone)]
-#[repr(C)]
-pub struct CdfThreadContext {
-    pub r#ref: *mut Rav1dRef,
-    pub data: CdfThreadContext_data,
-    pub progress: *mut AtomicU32,
+pub enum CdfThreadContext {
+    QCat(c_uint),
+    Cdf(Arc<CdfThreadContext_data>),
+}
+
+impl CdfThreadContext {
+    pub fn progress(&self) -> Option<&AtomicU32> {
+        match self {
+            Self::QCat(_) => None,
+            Self::Cdf(cdf) => cdf.progress.as_ref(),
+        }
+    }
+
+    pub fn cdf(&self) -> Option<&RwLock<CdfContext>> {
+        match self {
+            Self::QCat(_) => None,
+            Self::Cdf(cdf) => Some(&cdf.cdf),
+        }
+    }
+}
+
+impl Default for CdfThreadContext {
+    fn default() -> Self {
+        Self::QCat(0)
```

Comment:
nit: There's nothing special about 0 here, we just want some default value, so I think `Default::default()` expresses that intent slightly better.  Not a big deal, though.

```suggestion
        Self::QCat(Default::default())
```

Reply:
There is something special about zero, the original C code zero-initialized this field. I'd prefer to make that explicit here.

Reply:
Oh, there is?  Then your way is better.